lxd/instance/drivers/driver_qemu.go

```diff
@@ -762,18 +762,27 @@ func (d *qemu) Shutdown(timeout time.Duration) error {
 		return err
 	}
 
-	// If timeout provided, block until the VM is not running or the timeout has elapsed.
+	var timeoutCh <-chan time.Time // If no timeout specified, will be nil, and a nil channel always blocks.
 	if timeout > 0 {
+		timeoutCh = time.After(timeout)
+	}
+
+	for {
 		select {
 		case <-chDisconnect:
-			break
-		case <-time.After(timeout):
+			// VM monitor disconnected, VM is on the way to stopping, now wait for onStop() to finish.
+		case <-timeoutCh:
+			// User specified timeout has elapsed without VM stopping.
 			err = fmt.Errorf("Instance was not shutdown after timeout")
 			op.Done(err)
-			return err
+		case <-time.After((operationlock.TimeoutSeconds / 2) * time.Second):
+			// Keep the operation alive so its around for onStop() if the VM takes
+			// longer than the default 30s that the operation is kept alive for.
+			op.Reset()
+			continue
```

Comment:
This worries me a bit as it means that the operation will never expire, every 15s, we reset it for another 30s but there's no way for the user to abort it and if the user issues a new `lxc stop`, it will just wait on the existing operation and won't attempt to send the reboot signal again.

Reply:
Should the `lxc stop` operation itself be cancellable? Currently it is not, so doing a ctrl+c on a `lxc stop` command that is hanging leave the AP endpoint go routine running until the VM stops.

Reply:
I don't think we want to open that can of worms right now (make those cancellable) but if the user ctrl+c the CLI and issues `lxc stop` again (potentially with a different timeout), we don't want it to get stuck indefinitely on the existing operation.

See below for expected behavior, hopefully that helps.